altos-rust/port/cortex-m0/src/lib.rs

```diff
@@ -41,172 +44,184 @@ pub use exceptions::EXCEPTIONS;
-  }
-
-  pub mod sync {
-    pub use altos_core::sync::{Mutex, MutexGuard};
-    pub use altos_core::sync::CondVar;
-    pub use altos_core::sync::CriticalSection;
-  }
+    pub mod alloc {
+        pub use altos_core::alloc::boxed::Box;
+    }
+
+    pub mod collections {
+        // TODO: Do we want to expose an allocation interface?
+        pub use altos_core::collections::Vec;
+        pub use altos_core::queue::{SortedList, Queue, Node};
+    }
+
+    pub mod sync {
+        pub use altos_core::sync::{Mutex, MutexGuard};
+        pub use altos_core::sync::CondVar;
+        pub use altos_core::sync::CriticalSection;
+    }
 }
 
-#[cfg(not(test))]
+#[cfg(not(any(test, feature="doc")))]
 #[lang = "eh_personality"] extern "C" fn eh_personality() {}
-#[cfg(not(test))]
+#[cfg(not(any(test, feature="doc")))]
 #[lang = "panic_fmt"]
-extern "C" fn panic_fmt(_fmt: core::fmt::Arguments, _file: &'static str, _line: usize) -> ! {
-  loop {
-    unsafe {
-      arm::asm::bkpt();
+extern "C" fn panic_fmt(fmt: core::fmt::Arguments,
+                        (file, line): (&'static str, u32)) -> ! {
+    unsafe { arm::asm::disable_interrupts() };
+    kprintln!("Panicked at File: {}, Line: {}", file, line);
+    kprintln!("{}", fmt);
+    loop {
+        unsafe {
+            arm::asm::bkpt();
+        }
     }
-  }
 }
 
 extern {
-  // The application layer's entry point
-  fn application_entry() -> !;
+    // The application layer's entry point
+    fn application_entry() -> !;
 }
 
 #[export_name="_reset"]
 pub fn init() -> ! {
-  // TODO: set pendsv and systick interrupts to lowest priority
-  unsafe { arm::asm::disable_interrupts() };
-  init_data_segment();
-  init_bss_segment();
-  init_heap();
-  init_led();
-  init_clock();
-  init_ticks();
-
-  unsafe { application_entry() };
+    // TODO: set pendsv and systick interrupts to lowest priority
+    unsafe { arm::asm::disable_interrupts() };
+    init_data_segment();
+    init_bss_segment();
+    init_heap();
+    init_led();
+    init_clock();
+    init_ticks();
+    init_usart();
+
+    kprintln!("Initialization Complete");
+
+    unsafe { application_entry() };
 }
 
 fn init_data_segment() {
-  #[cfg(target_arch="arm")]
-  unsafe {
-    asm!(
-      concat!(
-        "ldr r1, =_sidata\n", /* start of data in flash */
-        "ldr r2, =_sdata\n",  /* start of memory location in RAM */
-        "ldr r3, =_edata\n",  /* end of memory location in RAM */
-      "copy:\n",
-        "cmp r2, r3\n", /* check if we've reached the end of our segment */
-        "bpl d_done\n",
-        "ldr r0, [r1]\n", /* if not, keep copying */
-        "adds r1, #4\n",
-        "str r0, [r2]\n",
-        "adds r2, #4\n",
-        "b copy\n", /* repeat until done */
-      "d_done:\n")
-    : /* no outputs */
-    : /* no inputs */
-    : "r0", "r1", "r2", "r3" /* clobbers */
-    : "volatile");
-  }
+#[cfg(target_arch="arm")]
+    unsafe {
+        asm!(
+                concat!(
```

Comment:
All the strings are technically inside the `concat!` macro, this just makes it easier to comment lines